File: Data Structures & Algorithms/split-array-largest-sum/submission-1.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int max = 0;
        int total = 0;
        for (int n: nums) {
            max = Math.max(max, n);
            total += n;
        }

        int l = max;
        int r = total;
        int res = total;

        while (l < r) {
            int m = l + ((r - l)/2);
            if (canSplit(nums, m, k)) {
                res = m;
                r = m;
            } else {
                l = m + 1;
            }
        }

        return res;
    }

    private static boolean canSplit(final int[] nums, final int expectedSum, final int k) {
        int currSum = 0;
        int splits = 1;
        for (int n: nums) {
            if (currSum + n > expectedSum) {
                ++splits;
                currSum = 0;
            }
            currSum += n;
        }
        return splits <= k;
    }
}

File: Data Structures & Algorithms/subarray-sums-divisible-by-k/submission-1.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        prefix_mods = [0] * k
        prefix_mods[0] = 1
        curr_prefix_mod = 0
        res = 0
        for n in nums:
            curr_prefix_mod = (curr_prefix_mod + n % k + k) % k
            res += prefix_mods[curr_prefix_mod]
            prefix_mods[curr_prefix_mod] += 1
        return res 